qutebrowser/javascript/greasemonkey_wrapper.js

```diff
@@ -145,9 +145,63 @@
         }
     };
 
-    const unsafeWindow = window;
-
-    // ====== The actual user script source ====== //
+    {% if use_proxy %}
+      /*
+       * Try to give userscripts an environment that they expect. Which
+       * seems to be that the global window object should look the same as
+       * the page's one and that if a script writes to an attribute of
+       * window it should be able to access that variable in the global
+       * scope.
+       * Use a Proxy to stop scripts from actually changing the global
+       * window (that's what unsafeWindow is for).
+       * Use the "with" statement to make the proxy provide what looks
+       * like global scope.
+       *
+       * There are other Proxy functions that we may need to override.
+       * set, get and has are definitely required.
+       */
+      const unsafeWindow = window;
+      const qute_gm_window_shadow = {};  // stores local changes to window
+      const qute_gm_windowProxyHandler = {
+        get: function(obj, prop) {
+          if (prop in qute_gm_window_shadow)
+            return qute_gm_window_shadow[prop];
+          if (prop in obj) {
+            if (typeof obj[prop] === 'function' && typeof obj[prop].prototype == 'undefined')
+              // Getting TypeError: Illegal Execution when callers try to execute
+              // eg addEventListener from here because they were returned
+              // unbound
+              return obj[prop].bind(obj);
+            return obj[prop];
+          }
+        },
+        set: function(target, prop, val) {
+          return qute_gm_window_shadow[prop] = val;
+        },
+        has: function(target, key) {
+          return key in qute_gm_window_shadow ? true :
+                 key in target;
+        }
+      };
+      const qute_gm_window_proxy = new Proxy(
+        unsafeWindow, qute_gm_windowProxyHandler);
+
+      with (qute_gm_window_proxy) {
+        // We can't return `this` or `qute_gm_window_proxy` from
+        // `qute_gm_window_proxy.get('window')` because the Proxy implementation
+        // does typechecking on read-only things. So we have to shadow `window`
+        // more conventionally here. Except we can't do it directly within
+        // with `with` scope because then it would get assigned to the
+        // proxy and we would get the same problem, so we have to make yet
+        // another nested scope.
```

Comment:
I don't get why the function-IIFE is needed. Shouldn't `let` take care of making it block-scoped (with the userscript) instead of assigning it to the proxy?

However, the tests also don't seem to catch this - if I remove the function, they don't fail. If I also remove the `let`, they still don't fail :wink:

Reply:
Thank you for being so thorough, I guess I was doing something silly (not using a keyword?) and then did that to code around that and then later stopped doing the first silly thing. Everything appears to be fine without it, I also changed the `let` to a `const` because why not?